go/pkg/contextmd/contextmd.go

```diff
@@ -177,33 +177,33 @@ func mergeSet(set map[string]struct{}) string {
 
 // capToLimit ensures total length does not exceed max header size.
 func capToLimit(m *Metadata, limit int) *Metadata {
-	total := len(m.ToolName) + len(m.ToolVersion) + len(m.ActionID) + len(m.InvocationID) + len(m.CorrelatedInvocationID)
+	total := len(m.ToolName) + len(m.ToolVersion) + len(m.ActionID) + len(m.InvocationID) + len(m.CorrelatedInvocationsID)
 	excess := total - limit
 	if excess <= 0 {
 		return m
 	}
-	// We ignore the tool name, because in practice this is a
-	// very short constant which makes no sense to truncate.
-	diff := len(m.ActionID) - len(m.InvocationID)
-	if diff > 0 {
-		if diff > excess {
-			m.ActionID = m.ActionID[:len(m.ActionID)-excess]
-		} else {
-			m.ActionID = m.ActionID[:len(m.ActionID)-diff]
-			rem := (excess - diff + 1) / 2
-			m.ActionID = m.ActionID[:len(m.ActionID)-rem]
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-rem]
-		}
-	} else {
-		diff = -diff
-		if diff > excess {
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-excess]
-		} else {
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-diff]
-			rem := (excess - diff + 1) / 2
-			m.InvocationID = m.InvocationID[:len(m.InvocationID)-rem]
-			m.ActionID = m.ActionID[:len(m.ActionID)-rem]
+	truncateStrings(limit, &m.ActionID, &m.InvocationID, &m.CorrelatedInvocationsID, &m.ToolName, &m.ToolVersion)
+	return m
+}
+
+// truncateStrings remove one char from the longest str at a time until the
+// total length of all the strings is less than limit value.
+func truncateStrings(limit int, inputs ...*string) {
```

Comment:
Any reason to extract the code to this helper function?

Reply:
I have a helper function here because I think 

1) how we truncate strings (say, maybe we want to, if possible, only remove the entire action_id, without touching the invocation_id because  invocation_id seems more important, and action_digest is import, but no one works with action_id), and

 2) what strings need to be truncated, and what strings we will not touch (say, maybe the tool version, we might don't want it to be 0.15[missing digits]),
 
 might subject to change....

I don't have much context about why and in which case we will have this header limit overflow problem, and I am not sure what side effect will be after we truncated the strings....

I think lets have a helper function here for now, and until we see a real break, we will come back here and refactor the truncateString logic.

Basically, the parent function indicates we must truncate String when it overflow (this part is solid)
the helper function decides which strategy we use to truncate strings (this we don't have enough info to make decision now, and might subject to change, so I move this part out.)

Reply:
Ack